File: src/crewlet/mattermost/doctor.py
"""``crewlet mattermost doctor`` — prove a Mattermost install actually works.

Mirrors ``crewlet llm doctor``'s job for the other backend an operator
has to stand up themselves: check the things that are *silently* wrong,
not the things that raise.

Two of those dominate, and neither shows up as an error anywhere:

**The Site URL.** ``ServiceSettings.SiteURL`` is not cosmetic. Mattermost
accepts a websocket upgrade only from a browser whose ``Origin`` matches
it *exactly* — host and scheme — so a server that answers every REST call
correctly on ``http://10.0.0.7:8065`` will refuse the web app's live
connection when its Site URL still says ``http://localhost:8065``. The
web app then falls back to fetching on navigation, which reads as "I have
to refresh to see replies" rather than as an outage. Plugins compound it:
they build their URLs from the same value, so their requests go to a host
the reader's own machine has to resolve. The engine is *exempt* — a
non-browser client sends no ``Origin`` header and Mattermost lets those
through — which is exactly why this can run for weeks: the agents answer,
and only the humans' clients are blind.

**A seat's websocket.** A bot token can be syntactically fine, belong to
a real account, and still not open a socket — the bot was disabled, the
token was revoked, personal access tokens were turned off server-wide.
The engine's *only* inbound path is that socket, so the failure mode is
an agent that never hears anything while every other check passes. So
this command opens the real handshake and authenticates, the way
``llm doctor`` insists on a real completion.
"""

from __future__ import annotations

import asyncio
from dataclasses import dataclass, field

from crewlet._logging import get_logger
from crewlet.mattermost.client import (
    MattermostClient,
    MattermostError,
    normalize_base_url,
    site_urls_match,
    websocket_url,
)

logger = get_logger("mattermost.doctor")

#: How long the browser-shaped upgrade probe and each seat's
#: authentication round trip may take.  Both are a single round trip on a
#: server that has already answered ``/system/ping``, so this only has to
#: survive a loaded server; it is deliberately short enough that a doctor
#: run against an unreachable host fails in seconds rather than minutes.
PROBE_TIMEOUT_SECONDS = 15.0


@dataclass
class SeatCheck:
    """One agent seat's Mattermost health."""

    handle: str
    username: str = ""
    user_id: str = ""
    token_ok: bool = False
    is_bot: bool = False
    disabled: bool = False
    websocket_ok: bool = False
    channels: list[str] = field(default_factory=list)
    problems: list[str] = field(default_factory=list)

    @property
    def ok(self) -> bool:
        return not self.problems


@dataclass
class DoctorReport:
    """Everything ``crewlet mattermost doctor`` learned about one install."""

    url: str
    team: str
    reachable: bool = False
    server_version: str = ""
    site_url: str = ""
    site_url_ok: bool = False
    browser_websocket_ok: bool = False
    browser_websocket_detail: str = ""
    team_ok: bool = False
    seats: list[SeatCheck] = field(default_factory=list)
    problems: list[str] = field(default_factory=list)

    @property
    def ok(self) -> bool:
        return not self.problems and all(seat.ok for seat in self.seats)


async def _probe_browser_websocket(base_url: str) -> tuple[bool, str]:
    """Open the websocket the way a BROWSER would, and report the outcome.

    The ``origin`` argument is the whole point: it is the one header a
    browser always sends and a server-side client never does, and it is
    the only input to Mattermost's websocket origin check.  Probing
    without it would prove the endpoint is reachable while telling us
    nothing about whether a human's web app can use it — which is
    precisely the blind spot this command exists to remove.
    """
    import websockets

    url = websocket_url(base_url)
    origin = normalize_base_url(base_url)
    try:
        async with asyncio.timeout(PROBE_TIMEOUT_SECONDS):
            async with websockets.connect(url, origin=origin, open_timeout=None):
                return True, f"upgraded with Origin: {origin}"
    except TimeoutError:
        return False, f"timed out after {PROBE_TIMEOUT_SECONDS:.0f}s"
    except Exception as exc:  # websockets raises a family of these
        return False, f"{type(exc).__name__}: {exc}"


async def _probe_seat_websocket(base_url: str, token: str, handle: str) -> str:
    """Authenticate one seat's websocket; ``""`` when it works.

    No ``origin``: this is the connection the ENGINE makes, and it must
    be checked as the engine makes it.  Sending one here would fold the
    Site URL problem into every seat's result and hide whichever of the
    two is actually broken.
    """
    import websockets

    from crewlet.mattermost.events import (
        MattermostAuthError,
        await_authentication,
        send_authentication_challenge,
    )

    try:
        async with asyncio.timeout(PROBE_TIMEOUT_SECONDS):
            async with websockets.connect(
                websocket_url(base_url), open_timeout=None
            ) as socket:
                await send_authentication_challenge(socket, token)
                await await_authentication(
                    socket, handle=handle, timeout=PROBE_TIMEOUT_SECONDS
                )
        return ""
    except MattermostAuthError as exc:
        return str(exc)
    except TimeoutError:
        return f"no authentication reply within {PROBE_TIMEOUT_SECONDS:.0f}s"
    except Exception as exc:
        return f"{type(exc).__name__}: {exc}"


async def _check_seat(
    base_url: str,
    team_id: str,
    handle: str,
    token: str,
) -> SeatCheck:
    """Token → identity → membership → a real authenticated websocket."""
    seat = SeatCheck(handle=handle)
    if not token:
        seat.problems.append(
            "no bot token resolved — the ${VAR} in "
            "role.integrations.mattermost.bot_token is unset; run "
            "`crewlet mattermost provision`"
        )
        return seat

    client = MattermostClient(base_url, token)
    try:
        try:
            me = await client.me()
        except MattermostError as exc:
            seat.problems.append(
                f"the bot token is not usable ({exc}) — it may have been "
                "revoked, or its bot account disabled; re-run "
                "`crewlet mattermost provision`"
            )
            return seat
        seat.token_ok = True
        seat.user_id = str((me or {}).get("id") or "")
        seat.username = str((me or {}).get("username") or "")
        seat.is_bot = bool((me or {}).get("is_bot"))
        seat.disabled = bool((me or {}).get("delete_at"))
        if seat.disabled:
            seat.problems.append(
                "the account is disabled — a later provision run re-enables it"
            )

        if team_id and seat.user_id:
            try:
                channels = await client.list_channels_for_user(seat.user_id, team_id)
            except MattermostError as exc:
                seat.problems.append(f"cannot list its channels: {exc}")
            else:
                seat.channels = sorted(
                    str(c.get("name") or "") for c in channels if c.get("name")
                )
                if not seat.channels:
                    seat.problems.append(
                        "member of no channel in this team — a bot only "
                        "receives messages from channels it has joined, so "
                        "it would never be woken by anything but a DM"
                    )

        failure = await _probe_seat_websocket(base_url, token, handle)
        seat.websocket_ok = not failure
        if failure:
            seat.problems.append(f"websocket authentication failed: {failure}")
    finally:
        await client.close()
    return seat


async def run_doctor(
    *,
    url: str,
    team: str,
    seat_tokens: dict[str, str],
) -> DoctorReport:
    """Check one Mattermost install and every seat configured against it."""
    base_url = normalize_base_url(url)
    report = DoctorReport(url=base_url, team=team)

    # Unauthenticated on purpose: /system/ping and /config/client are the
    # two reads a browser makes before anyone signs in, so a bad operator
    # credential cannot turn a healthy server into a red report.
    anon = MattermostClient(base_url, "")
    try:
        try:
            ping = await anon.ping()
        except MattermostError as exc:
            report.problems.append(
                f"cannot reach {base_url}: {exc} — check the URL, that the "
                "server is running, and that this host can route to it"
            )
            return report
        report.reachable = True
        report.server_version = str((ping or {}).get("version") or "")

        config = await anon.client_config()
        report.site_url = normalize_base_url(str(config.get("SiteURL") or ""))
        report.site_url_ok = bool(report.site_url) and site_urls_match(
            base_url, report.site_url
        )
        if not report.site_url:
            report.problems.append(
                "the server did not report a SiteURL — every browser-facing "
                "URL it builds will be wrong"
            )
        elif not report.site_url_ok:
            report.problems.append(
                f"the server's SiteURL is {report.site_url}, not {base_url}. "
                "Mattermost accepts a websocket upgrade only from a browser "
                "whose Origin matches SiteURL exactly (host and scheme), so "
                "everyone browsing at another address loses live updates and "
                "has to refresh to see new messages, and plugins request "
                f"{report.site_url} from the reader's own machine. Set "
                "MM_SERVICESETTINGS_SITEURL (compose: MATTERMOST_PUBLIC_URL) "
                "to the address people type, and recreate the container — "
                "environment-managed settings cannot be changed from the "
                "System Console or the API"
            )

        ok, detail = await _probe_browser_websocket(base_url)
        report.browser_websocket_ok = ok
        report.browser_websocket_detail = detail
        if not ok:
            report.problems.append(
                f"a browser-shaped websocket to {base_url} did not open "
                f"({detail}) — the web app's live feed is dead for everyone "
                "until this works"
            )
    finally:
        await anon.close()

    # The team lookup needs a credential; borrow the first seat's, since
    # every seat is a member of it and no operator token is required for
    # a command whose whole point is to be safe to run.
    team_id = ""
    for handle, token in seat_tokens.items():
        if not token:
            continue
        client = MattermostClient(base_url, token)
        try:
            found = await client.get_team_by_name(team)
            team_id = str((found or {}).get("id") or "")
        except MattermostError as exc:
            logger.debug("team_lookup_failed", handle=handle, error=str(exc))
        finally:
            await client.close()
        if team_id:
            break
    report.team_ok = bool(team_id)
    if not team_id and seat_tokens:
        report.problems.append(
            f"team {team!r} was not found, or no seat can see it — channels "
            "are team-scoped, so nothing can be delivered"
        )

    for handle in sorted(seat_tokens):
        report.seats.append(
            await _check_seat(base_url, team_id, handle, seat_tokens[handle])
        )
    if not seat_tokens:
        report.problems.append(
            "no Mattermost-enabled agent seats in this config — add "
            "role.integrations.mattermost.bot_token to at least one role"
        )
    return report


def format_report(report: DoctorReport) -> str:
    """Render a :class:`DoctorReport` for the terminal."""
    lines = [
        f"url           : {report.url}",
        f"reachable     : {'yes' if report.reachable else 'NO'}"
        + (f" (Mattermost {report.server_version})" if report.server_version else ""),
        f"site url      : {report.site_url or '(unreported)'}"
        + ("" if report.site_url_ok else "   << does not match url"),
        f"browser ws    : {'ok' if report.browser_websocket_ok else 'FAILED'}"
        + (
            f" — {report.browser_websocket_detail}"
            if report.browser_websocket_detail
            else ""
        ),
        f"team          : {report.team}{'' if report.team_ok else '   << not found'}",
    ]
    if report.seats:
        lines.append("")
        lines.append(f"{'SEAT':<18} {'USERNAME':<22} {'TOKEN':<7} {'WS':<7} CHANNELS")
        for seat in report.seats:
            lines.append(
                f"{seat.handle:<18} {seat.username or '-':<22} "
                f"{'ok' if seat.token_ok else 'FAILED':<7} "
                f"{'ok' if seat.websocket_ok else 'FAILED':<7} "
                f"{','.join(seat.channels) or '-'}"
            )
            for problem in seat.problems:
                lines.append(f"    - {problem}")
    if report.problems:
        lines.append("")
        lines.append("problems:")
        lines.extend(f"  - {p}" for p in report.problems)
    else:
        lines.append("")
        lines.append("problems      : none")
    return "\n".join(lines)


__all__ = [
    "PROBE_TIMEOUT_SECONDS",
    "DoctorReport",
    "SeatCheck",
    "format_report",
    "run_doctor",
]

File: internal/mattermost/doctor.go
package mattermost

import (
	"context"
	"errors"
	"fmt"
	"net/http"
	"sort"
	"strings"
	"time"

	"github.com/coder/websocket"

	"github.com/crewlet/crewlet/internal/config"
	"github.com/crewlet/crewlet/internal/org"
	"github.com/crewlet/crewlet/internal/provision"
)

// The health check for the things that fail SILENTLY.
//
// # Why this command exists at all
//
// Everything that fails loudly on Mattermost is already reported: a bad
// token 401s, a missing team 404s, a channel that is not there is a note in
// the provisioning report. What has no symptom is the Site URL.
//
// Mattermost accepts a websocket only from a client whose Origin matches
// ServiceSettings.SiteURL. The engine sends the Origin a browser at the
// CONFIGURED url would send — so when the two disagree, agents keep working
// (their sockets are refused, they reconnect, and the REST side is fine)
// while every human's browser silently loses its live feed. Nobody files a
// bug about a chat client that just seems slow.
//
// # So it dials, rather than only comparing strings
//
// A string comparison catches the common case and gets the rest wrong: a
// reverse proxy, a subpath deployment, an allowed-origins override. The
// only answer that means anything is a real upgrade, shaped exactly as a
// browser's, against the real server — and then a real authenticated socket
// per seat, because "the server accepts sockets" and "this bot's token
// opens one" are different questions and only the second one wakes an agent.

// Finding is one thing the doctor checked.
type Finding struct {
	// Check names what was asked, for the report.
	Check string
	// OK is whether it passed.
	OK bool
	// Detail says what was found, and — when it failed — what to do.
	Detail string
	// Fatal marks a failure that STOPPED the checks after it, because
	// everything downstream would have reported a consequence rather
	// than a cause.
	//
	// Visible in the report, and that is the point: one failing line
	// with nothing after it reads as "one thing is wrong", when what it
	// means is "nothing else was even asked".
	Fatal bool
}

// Report is a whole run.
type Report struct{ Findings []Finding }

// Healthy reports a run with nothing wrong.
func (r *Report) Healthy() bool {
	for _, f := range r.Findings {
		if !f.OK {
			return false
		}
	}
	return true
}

// Stopped reports a run that gave up before checking everything.
func (r *Report) Stopped() bool {
	for _, f := range r.Findings {
		if f.Fatal {
			return true
		}
	}
	return false
}

func (r *Report) add(check string, ok bool, format string, args ...any) {
	r.Findings = append(r.Findings, Finding{
		Check: check, OK: ok, Detail: fmt.Sprintf(format, args...),
	})
}

func (r *Report) fatal(check string, format string, args ...any) {
	r.Findings = append(r.Findings, Finding{
		Check: check, Detail: fmt.Sprintf(format, args...), Fatal: true,
	})
}

// Dialer opens a websocket the way the fleet does.
//
// Injected so the suite can drive the refusals a real server produces
// without one — an origin check is exactly the behaviour a fake HTTP server
// can model faithfully, and exactly the behaviour a mock cannot.
type Dialer func(ctx context.Context, target, origin, token string) error

// DoctorOptions are one run's inputs.
type DoctorOptions struct {
	// Client talks to the instance. Its token may be empty: the checks
	// that need one borrow a SEAT's, because those are the credentials
	// the engine actually uses and an operator should not have to mint
	// anything to ask whether their company works.
	Client *Client

	// Config is the company's mattermost block, resolved.
	Config *config.Mattermost

	// Org is the company, for the per-seat socket checks.
	Org *org.Organization

	// SeatToken resolves a seat's bot token. A seat whose token is
	// missing is reported rather than skipped: it is an agent that will
	// never wake, which is the whole question this command answers.
	SeatToken func(seat *org.Role) (string, bool)

	// Dial opens a socket. Nil takes the real one.
	Dial Dialer
}

// Doctor runs the checks, in the order things break.
func Doctor(ctx context.Context, opts DoctorOptions) (*Report, error) {
	if opts.Client == nil {
		return nil, errors.New("mattermost: no client")
	}
	if opts.Config == nil || !opts.Config.Enabled {
		return nil, errors.New("mattermost: the company config does not enable mattermost")
	}
	dial := opts.Dial
	if dial == nil {
		dial = RealDial
	}
	report := &Report{}

	// FIRST AND UNAUTHENTICATED. A bad credential must not make a healthy
	// server look dead: the two have completely different remedies, and
	// an operator sent to restart their instance over a revoked token has
	// been actively misled.
	//
	// This inherits the client's retry budget, so a server mid-restart is
	// waited out rather than reported as down — which is slower on a
	// genuinely dead host and correct, because a doctor that called a
	// restarting server dead would disagree with the engine about the one
	// word they both use.
	if err := opts.Client.Ping(ctx); err != nil {
		report.fatal("reachable",
			"%s did not answer: %v — nothing below could be asked", opts.Client.URL(), err)
		return report, nil
	}
	report.add("reachable", true, "%s answers", opts.Client.URL())

	// ALSO UNAUTHENTICATED: the client configuration is public, and the
	// one setting whose failure has no error message is in it.
	conf, err := opts.Client.ClientConfig(ctx)
	if err != nil {
		report.fatal("site url",
			"could not read the server's own configuration: %v — the one "+
				"setting whose failure has no error message is in it", err)
		return report, nil
	}
	checkSiteURL(report, opts.Config.URL, conf)

	reader, seats := credentialed(ctx, report, opts)
	if reader == nil {
		// EVERYTHING BELOW NEEDS A TOKEN, and none resolved. Reporting
		// each of them as a separate failure would send an operator
		// after faults nobody observed.
		return report, nil
	}
	team := checkTeam(ctx, report, opts, reader)
	checkBrowserSocket(ctx, report, opts, dial, reader.Token())
	checkSeats(ctx, report, opts, dial, seats, team)
	return report, nil
}

// credentialed picks the client the authenticated checks run as, and
// resolves every seat's own credential once.
//
// # No operator credential is required
//
// The seat tokens in the config are the credentials the engine authenticates
// with, so they are also the honest thing to check with — and asking an
// operator to mint an admin token to find out whether their company works is
// a step that exists only to be skipped. One is used if given.
func credentialed(ctx context.Context, report *Report, opts DoctorOptions) (*Client, []seatCredential) {
	seats := resolveSeats(opts)
	reader := opts.Client
	if strings.TrimSpace(reader.Token()) == "" {
		reader = nil
		for _, seat := range seats {
			if seat.token != "" {
				reader = seat.client(opts.Client)
				break
			}
		}
	}
	if reader == nil {
		report.fatal("credential",
			"no credential resolved: neither an operator token nor any seat's "+
				"bot token. Pass -admin-token, or make this deployment's "+
				"environment carry the variables the seats' mattermost."+
				"bot_token fields reference")
		return nil, seats
	}
	who, err := reader.Me(ctx)
	if err != nil {
		report.fatal("credential",
			"the credential the checks run as does not authenticate: %v — "+
				"nothing below could be asked", err)
		return nil, seats
	}
	report.add("credential", true, "authenticates as %s", who.Username)
	return reader, seats
}

// seatCredential is one agent seat and the credential it would use.
type seatCredential struct {
	handle string
	token  string
}

// client builds a client authenticating as this seat.
func (s seatCredential) client(base *Client) *Client {
	out, err := NewClient(ClientOptions{
		URL: base.URL(), Token: s.token, HTTP: base.HTTP(),
	})
	if err != nil {
		return nil
	}
	return out
}

// resolveSeats reads every chat seat's credential once.
func resolveSeats(opts DoctorOptions) []seatCredential {
	if opts.Org == nil {
		return nil
	}
	var out []seatCredential
	for _, seat := range chatSeats(opts.Org) {
		credential := seatCredential{handle: seat.Handle()}
		if opts.SeatToken != nil {
			if token, ok := opts.SeatToken(seat); ok {
				credential.token = strings.TrimSpace(token)
			}
		}
		out = append(out, credential)
	}
	return out
}

// checkSiteURL compares what the server thinks it is against what the
// company config says.
//
// TWO FINDINGS, not one, because the consequences differ and so do the
// fixes: an origin mismatch blinds every browser's live feed, while a path
// mismatch breaks the absolute links the server builds and leaves the
// socket working. Reporting them together would send an operator to fix a
// socket that is fine.
func checkSiteURL(report *Report, configured string, conf map[string]string) {
	reported := SiteURL(conf)
	if reported == "" {
		report.add("site url", false,
			"ServiceSettings.SiteURL is unset. Mattermost refuses every "+
				"websocket whose Origin does not match it, so with no value "+
				"set every browser loses its live feed while agents keep "+
				"working. Set it in System Console → Environment → Web Server")
		return
	}
	switch {
	case !OriginMatches(configured, reported):
		report.add("site url", false,
			"the server reports SiteURL %s and this company is configured "+
				"for %s. Mattermost accepts a websocket only from an Origin "+
				"matching its SiteURL, so every human's live feed fails while "+
				"agents keep working — the failure has no symptom anyone "+
				"reports. Make the two agree", reported, NormalizeURL(configured))
	case !PathMatches(configured, reported):
		report.add("site url", false,
			"the server reports SiteURL %s and this company is configured "+
				"for %s. The origin matches, so websockets work — but the "+
				"paths differ, and the server builds its absolute links and "+
				"plugin URLs from its own value",
			reported, NormalizeURL(configured))
	default:
		report.add("site url", true, "the server agrees it is served at %s", reported)
	}
}

// checkTeam resolves the configured team.
//
// A bot in no team receives nothing, and the provisioner's own report is
// the only other place this shows up — so an operator debugging silence
// after the fact needs it here. The resolved team is also what the per-seat
// channel check hangs off, which is why it comes back.
func checkTeam(ctx context.Context, report *Report, opts DoctorOptions, reader *Client) string {
	team, found, err := reader.TeamByName(ctx, opts.Config.Team)
	switch {
	case err != nil:
		report.add("team", false, "could not resolve team %q: %v", opts.Config.Team, err)
	case !found:
		report.add("team", false,
			"this server has no team %q. Channels are team-scoped, so no bot "+
				"can be placed and no agent will ever receive a message",
			opts.Config.Team)
	default:
		report.add("team", true, "team %q resolves (%s)", opts.Config.Team, team.ID)
		return team.ID
	}
	return ""
}

// checkBrowserSocket proves the upgrade a browser performs.
//
// SHAPED EXACTLY AS A BROWSER'S — a credential and the Origin a browser at
// the configured url would send — because that is the request whose refusal
// nobody sees. A probe that sent no Origin, or one carrying a path, would
// pass a check every real browser fails.
func checkBrowserSocket(ctx context.Context, report *Report, opts DoctorOptions, dial Dialer, token string) {
	origin := BrowserOrigin(opts.Config.URL)
	if origin == "" {
		report.add("browser socket", false,
			"%q is not a usable base URL, so no Origin can be derived from it",
			opts.Config.URL)
		return
	}
	target := WebsocketURL(opts.Config.URL)
	if err := dial(ctx, target, origin, token); err != nil {
		report.add("browser socket", false,
			"a browser-shaped upgrade to %s with Origin %s was refused: %v. "+
				"This is what every human's live feed does, and its failure "+
				"is silent — the client simply stops updating",
			target, origin, err)
		return
	}
	report.add("browser socket", true,
		"a browser-shaped upgrade to %s was accepted", target)
}

// checkSeats asks, per agent seat, whether that seat would actually wake.
//
// # "The server accepts sockets" is not "this bot wakes"
//
// A seat fails on its own while every other check passes: a token that was
// revoked, a bot that was deactivated, a `${VAR}` that never made it into
// this deployment's environment, a bot in no channel. Each leaves ONE agent
// permanently deaf while the company looks healthy, so each seat is checked
// with ITS OWN credential.
//
// # A seat that fails early is not asked the later questions
//
// A seat whose token did not resolve never gets a socket dialled, and one
// whose credential is refused is never asked about its channels. Reporting
// those as separate failures would send an operator after faults nobody
// observed — the answer is already in the first line.
func checkSeats(ctx context.Context, report *Report, opts DoctorOptions, dial Dialer,
	seats []seatCredential, teamID string,
) {
	target := WebsocketURL(opts.Config.URL)
	origin := BrowserOrigin(opts.Config.URL)
	for _, seat := range seats {
		check := "seat " + seat.handle
		if seat.token == "" {
			report.add(check, false,
				"no bot token resolved for this seat, so it can never open a "+
					"socket and will never receive a message. Check that its "+
					"mattermost.bot_token variable is set in this deployment's "+
					"environment, and run `crewlet mattermost provision` if it "+
					"has never been minted")
			continue
		}
		client := seat.client(opts.Client)
		if client == nil {
			report.add(check, false, "this seat's credential could not be used")
			continue
		}
		who, err := client.Me(ctx)
		if err != nil {
			report.add(check, false,
				"this seat's own credential is refused: %v. Its agent cannot "+
					"do anything at all, while everything else about the "+
					"company looks healthy", err)
			continue
		}
		if err = dial(ctx, target, origin, seat.token); err != nil {
			report.add(check, false,
				"this seat authenticates but cannot open a socket: %v. A token "+
					"valid for REST can still fail here, and the socket is the "+
					"engine's only inbound path — the agent is deaf", err)
			continue
		}
		if teamID == "" {
			// The team did not resolve, so the channel question has no
			// answer. The socket check above is still worth having.
			report.add(check, true, "%s authenticates and opens a socket", who.Username)
			continue
		}
		channels, err := client.Channels(ctx, who.ID, teamID)
		switch {
		case err != nil:
			report.add(check, false,
				"%s authenticates and opens a socket, but its channels could "+
					"not be read: %v", who.Username, err)
		case len(channels) == 0:
			// A BOT HEARS ONLY WHAT ITS CHANNELS DELIVER. One in none
			// is an agent that wakes for direct messages and nothing
			// else — and its account looks perfectly healthy.
			report.add(check, false,
				"%s authenticates and opens a socket but has joined no channel, "+
					"so it will only ever hear direct messages. Name channels "+
					"under integrations.mattermost.provisioning or on the seat "+
					"itself, and run `crewlet mattermost provision`", who.Username)
		default:
			report.add(check, true, "%s authenticates, opens a socket, and is in %d channel(s)",
				who.Username, len(channels))
		}
	}
}

// chatSeats is every agent seat that declares a Mattermost bot, in a stable
// order.
func chatSeats(o *org.Organization) []*org.Role {
	var out []*org.Role
	for seat := range o.AllRoles() {
		if seat.IsAgent() && !seat.Mattermost.IsZero() {
			out = append(out, seat)
		}
	}
	sort.Slice(out, func(i, j int) bool { return out[i].Handle() < out[j].Handle() })
	return out
}

// SeatTokens resolves each seat's bot token through the environment, the
// way the engine does.
func SeatTokens(env *config.Resolver) func(*org.Role) (string, bool) {
	return func(seat *org.Role) (string, bool) {
		raw := seat.Mattermost.BotToken
		if strings.TrimSpace(raw) == "" {
			return "", false
		}
		// A LITERAL IS A VALUE, not a reference: an operator managing a
		// seat's credential by hand is a supported choice, and refusing
		// to check it would report a working seat as unconfigured.
		if _, ok := provision.SoleVar(raw); !ok && len(provision.ReferencedVars(raw)) == 0 {
			return raw, true
		}
		value := strings.TrimSpace(env.Value(raw))
		return value, value != ""
	}
}

// RealDial opens a websocket against the instance and closes it.
//
// The connection is made and dropped: what is being asked is whether the
// upgrade is accepted and the credential authenticates, and holding it open
// would only add ways for the check itself to fail.
func RealDial(ctx context.Context, target, origin, token string) error {
	ctx, cancel := context.WithTimeout(ctx, DialTimeout)
	defer cancel()
	header := http.Header{}
	if token != "" {
		header.Set("Authorization", "Bearer "+token)
	}
	if origin != "" {
		header.Set("Origin", origin)
	}
	return dialOnce(ctx, target, header)
}

// DialTimeout bounds one probe.
//
// SHORT. Every check here is a single round trip against an instance the
// operator is sitting in front of, and the failure this command exists to
// find is a refusal — which arrives immediately. A long timeout would turn
// a down server into a command that appears to hang.
const DialTimeout = 10 * time.Second

// dialOnce performs the upgrade and closes it.
func dialOnce(ctx context.Context, target string, header http.Header) error {
	//nolint:bodyclose // On a successful upgrade the library has already
	// closed the handshake response; on a failure there is no body to close.
	conn, _, err := websocket.Dial(ctx, target, &websocket.DialOptions{HTTPHeader: header})
	if err != nil {
		return err
	}
	return conn.Close(websocket.StatusNormalClosure, "")
}
